dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/icon/IconTest.java

```diff
@@ -0,0 +1,227 @@
+
+import lombok.SneakyThrows;
+
+import org.hisp.dhis.feedback.BadRequestException;
+import org.hisp.dhis.feedback.NotFoundException;
+import org.hisp.dhis.fileresource.FileResource;
+import org.hisp.dhis.fileresource.FileResourceDomain;
+import org.hisp.dhis.fileresource.FileResourceService;
+import org.hisp.dhis.tracker.TrackerTest;
+import org.hisp.dhis.user.CurrentUserService;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.util.MimeTypeUtils;
+
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hashing;
+
+class IconTest extends TrackerTest
+{
+    @Autowired
+    private FileResourceService fileResourceService;
+
+    @Autowired
+    private CurrentUserService currentUserService;
+
+    @Autowired
+    private IconService iconService;
+
+    private final String[] keywords = { "k1", "k2", "k3" };
+
+    @SneakyThrows
+    @Override
+    protected void initTest()
+        throws IOException
+    {
+        FileResource fileResource = createAndPersistFileResource( 'A' );
+        String iconKey = "iconKey";
+        iconService.addCustomIcon(
+            new CustomIcon( iconKey, "description", keywords, fileResource.getUid(),
+                currentUserService.getCurrentUser().getUid() ) );
+    }
+
+    @Test
+    void shouldGetAllIconsWhenRequested()
+    {
+        Map<String, DefaultIcon> defaultIconMap = getAllDefaultIcons();
+
+        assertEquals( defaultIconMap.size() + 1, iconService.getIcons().size(),
+            String.format( "Expected to find %d icons, but found %d instead", defaultIconMap.size() + 1,
+                iconService.getIcons().size() ) );
+    }
+
+    @Test
+    void shouldGetDefaultIconWhenKeyBelongsToDefaultIcon()
+        throws NotFoundException
+    {
+        String defaultIconKey = getAllDefaultIcons().keySet().stream().findAny().orElse( null );
+
+        Icon icon = iconService.getIcon( defaultIconKey );
+
+        assertEquals( defaultIconKey, icon.getKey() );
+    }
+
+    @Test
+    void shouldGetAllKeywordsWhenRequested()
+    {
+        Set<String> keywordList = getAllDefaultIcons().values().stream()
+            .map( Icon::getKeywords )
+            .flatMap( Arrays::stream ).collect( Collectors.toSet() );
+
+        assertEquals( keywordList.size() + keywords.length, iconService.getKeywords().size(),
+            String.format( "Expected to find %d icons, but found %d instead", keywordList.size() + keywords.length,
+                iconService.getIcons().size() ) );
+    }
+
+    @Test
+    void shouldGetIconsFilteredByKeywordWhenRequested()
+        throws BadRequestException,
+        NotFoundException
+    {
+        Optional<DefaultIcon> defaultIcon = getAllDefaultIcons().values().stream()
+            .filter( si -> si.getKeywords().length > 0 ).findAny();
+
+        if ( defaultIcon.isEmpty() )
+        {
+            return;
```

Comment:
Why would we not test getting icons by keyword if there is no default icon with keywords?